// ngx-i18nsupport/dist/xliffmerge/xlf20-merge.spec.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var fs = require("fs");
var xliff_merge_1 = require("./xliff-merge");
var command_output_1 = require("../common/command-output");
var writer_to_string_1 = require("../common/writer-to-string");
var file_util_1 = require("../common/file-util");
var translation_messages_file_reader_1 = require("./translation-messages-file-reader");
var dist_1 = require("ngx-i18nsupport-lib/dist");
var auto_translate_service_spec_1 = require("../autotranslate/auto-translate-service.spec");
var xml_reader_1 = require("./xml-reader");
/**
 * Created by martin on 18.02.2017.
 * Testcases for XliffMerge Format XLIFF 2.0.
 */
describe('XliffMerge XLIFF 2.0 format tests', function () {
    /**
     * Workdir, not in git.
     * Cleaned up for every test.
     * Tests, that work on files, copy everything they need into this directory.
     * @type {string}
     */
    var WORKDIR = 'test/work/';
    var SRCDIR = 'test/testdata/i18n/';
    var ENCODING = 'UTF-8';
    /**
     * Helper function to read XLIFF 2.0 from File
     * @type {string}
     */
    function readXliff2(path) {
        return translation_messages_file_reader_1.TranslationMessagesFileReader.fromFile('xlf2', path, ENCODING);
    }
    describe('Merge process checks for format XLIFF 2.0', function () {
        var MASTER1FILE = 'ngExtractedMaster1.xlf2';
        var MASTER2FILE = 'ngExtractedMaster2.xlf2';
        var MASTER1SRC = SRCDIR + MASTER1FILE;
        var MASTER2SRC = SRCDIR + MASTER2FILE;
        var MASTERFILE = 'messages.xlf2';
        var MASTER = WORKDIR + MASTERFILE;
        var ID_APP_RUNS = "4371668001355139802"; // an ID from ngExtractedMaster1.xlf
        var ID_REMOVED_MYFIRST = "2047558209369508311"; // an ID that will be removed in master2
        var ID_REMOVED_APPDESCRIPTION = "7499557905529977371"; // another removed ID
        var ID_WITH_PLACEHOLDER = "9030312858648510700";
        var ID_DESCRIPTION_CHANGE = 'DescriptionAndMeaning1';
        var ID_DESCRIPTION_ADD = 'AddDescriptionAndMeaning';
        var ID_DESCRIPTION_REMOVE = 'RemoveDescriptionAndMeaning';
        beforeEach(function () {
            if (!fs.existsSync(WORKDIR)) {
                fs.mkdirSync(WORKDIR);
            }
            // cleanup workdir
            file_util_1.FileUtil.deleteFolderContentRecursive(WORKDIR);
        });
        it('should fix source language, if the masters lang is not the default', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var master = readXliff2(MASTER);
            expect(master.sourceLanguage()).toBe('en'); // master is german, but ng-18n extracts it as en
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            expect(ws.writtenData()).toContain('master says to have source-language="en"');
            expect(ws.writtenData()).toContain('changed master source-language="en" to "de"');
            var newmaster = readXliff2(MASTER);
            expect(newmaster.sourceLanguage()).toBe('de'); // master is german
            done();
        });
        it('should generate translated file for default language de from master', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    useSourceAsTarget: false
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var langFile = readXliff2(xliffMergeCmd.generatedI18nFile('de'));
            expect(langFile.sourceLanguage()).toBe('de');
            expect(langFile.targetLanguage()).toBe('de');
            langFile.forEachTransUnit(function (tu) {
                expect(tu.targetContent()).toBe(tu.sourceContent());
                expect(tu.targetState()).toBe('final');
            });
            done();
        });
        it('should generate translated file for all languages', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var langFileGerman = readXliff2(xliffMergeCmd.generatedI18nFile('de'));
            expect(langFileGerman.sourceLanguage()).toBe('de');
            expect(langFileGerman.targetLanguage()).toBe('de');
            langFileGerman.forEachTransUnit(function (tu) {
                expect(tu.targetContent()).toBe(tu.sourceContent());
                expect(tu.targetState()).toBe('final');
            });
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            expect(langFileEnglish.sourceLanguage()).toBe('de');
            expect(langFileEnglish.targetLanguage()).toBe('en');
            langFileEnglish.forEachTransUnit(function (tu) {
                expect(tu.targetContent()).toBe(tu.sourceContent());
                expect(tu.targetState()).toBe('new');
            });
            done();
        });
        it('should generate translated file for all languages with empty targets for non default languages', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    useSourceAsTarget: false
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var langFileGerman = readXliff2(xliffMergeCmd.generatedI18nFile('de'));
            expect(langFileGerman.sourceLanguage()).toBe('de');
            expect(langFileGerman.targetLanguage()).toBe('de');
            langFileGerman.forEachTransUnit(function (tu) {
                expect(tu.targetContent()).toBe(tu.sourceContent());
                expect(tu.targetState()).toBe('final');
            });
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            expect(langFileEnglish.sourceLanguage()).toBe('de');
            expect(langFileEnglish.targetLanguage()).toBe('en');
            langFileEnglish.forEachTransUnit(function (tu) {
                expect(tu.targetContent()).toBe('');
                expect(tu.targetState()).toBe('new');
            });
            done();
        });
        it('should generate translated file for all languages with set praefix and suffix (#70)', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    targetPraefix: '%%',
                    targetSuffix: '!!',
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var langFileGerman = readXliff2(xliffMergeCmd.generatedI18nFile('de'));
            expect(langFileGerman.sourceLanguage()).toBe('de');
            expect(langFileGerman.targetLanguage()).toBe('de');
            langFileGerman.forEachTransUnit(function (tu) {
                if (!tu.targetContent().startsWith('{VAR')) {
                    expect(tu.targetContent()).toBe('%%' + tu.sourceContent() + '!!');
                }
                else {
                    expect(tu.targetContent()).toBe(tu.sourceContent());
                }
                expect(tu.targetState()).toBe('final');
            });
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            expect(langFileEnglish.sourceLanguage()).toBe('de');
            expect(langFileEnglish.targetLanguage()).toBe('en');
            langFileEnglish.forEachTransUnit(function (tu) {
                if (!tu.targetContent().startsWith('{VAR')) {
                    expect(tu.targetContent()).toBe('%%' + tu.sourceContent() + '!!');
                }
                else {
                    expect(tu.targetContent()).toBe(tu.sourceContent());
                }
                expect(tu.targetState()).toBe('new');
            });
            done();
        });
        it('should generate translated file with native trans unit status "initial", testcase for issue #57', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    useSourceAsTarget: false
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var langFileGerman = readXliff2(xliffMergeCmd.generatedI18nFile('de'));
            langFileGerman.forEachTransUnit(function (tu) {
                // need a cast to <any> to call nativeTargetState, which is not part of the official API
                expect(tu.nativeTargetState()).toBe('final');
            });
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            langFileEnglish.forEachTransUnit(function (tu) {
                // need a cast to <any> to call nativeTargetState, which is not part of the official API
                expect(tu.nativeTargetState()).toBe('initial'); // #56, state should be new, but in file we expect initial
            });
            done();
        });
        it('should merge translated file for all languages', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            // now translate some texts in the English version
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            var tu = langFileEnglish.transUnitWithId(ID_APP_RUNS);
            expect(tu).toBeTruthy();
            tu.translate('App runs');
            translation_messages_file_reader_1.TranslationMessagesFileReader.save(langFileEnglish);
            // next step, use another master
            file_util_1.FileUtil.copy(MASTER2SRC, MASTER);
            ws = new writer_to_string_1.WriterToString();
            commandOut = new command_output_1.CommandOutput(ws);
            xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            expect(ws.writtenData()).toContain('merged 1 trans-units from master to "en"');
            expect(ws.writtenData()).toContain('removed 2 unused trans-units in "en"');
            expect(ws.writtenData()).toContain('WARNING: transferred 1 source references');
            // look, that the new file contains the old translation
            langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            expect(langFileEnglish.transUnitWithId(ID_APP_RUNS).targetContent()).toBe('App runs');
            // look, that the removed IDs are really removed.
            expect(langFileEnglish.transUnitWithId(ID_REMOVED_MYFIRST)).toBeFalsy();
            expect(langFileEnglish.transUnitWithId(ID_REMOVED_APPDESCRIPTION)).toBeFalsy();
            done();
        });
        it('should merge changed source content with explicit ID to default language file (#81)', function (done) {
            var ID_SOURCE_CHANGE = 'sourcechanged';
            var ID_SOURCE_CHANGE_STATE_FINAL = 'sourcechanged_state_final';
            var TRANSLATED_FILE = 'WithSourceContentChange.de.xlf2';
            file_util_1.FileUtil.copy(SRCDIR + 'ngExtractedMasterWithSourceContentChange.xlf2', MASTER);
            file_util_1.FileUtil.copy(SRCDIR + TRANSLATED_FILE, WORKDIR + 'messages.de.xlf');
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'], verbose: true }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            expect(ws.writtenData()).toContain('WARNING: transferred 2 changed source content from master to "de"');
            // check that source is changed
            var langFileGerman = readXliff2(xliffMergeCmd.generatedI18nFile('de'));
            var tu = langFileGerman.transUnitWithId(ID_SOURCE_CHANGE);
            expect(tu).toBeTruthy();
            expect(tu.sourceContent()).toBe('Test Änderung Source (geändert!)');
            expect(tu.targetContent()).toBe('Test Änderung Source (geändert!)');
            expect(tu.targetState()).toBe(dist_1.STATE_FINAL);
            done();
        });
        it('should merge changed source content to already translated files', function (done) {
            var ID_SOURCE_CHANGE = 'sourcechanged';
            var ID_SOURCE_CHANGE_STATE_FINAL = 'sourcechanged_state_final';
            var TRANSLATED_FILE = 'WithSourceContentChange.en.xlf2';
            file_util_1.FileUtil.copy(SRCDIR + 'ngExtractedMasterWithSourceContentChange.xlf2', MASTER);
            file_util_1.FileUtil.copy(SRCDIR + TRANSLATED_FILE, WORKDIR + 'messages.en.xlf');
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'], verbose: true }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            expect(ws.writtenData()).toContain('WARNING: transferred 2 changed source content from master to "en"');
            // check that source is changed
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            var tu = langFileEnglish.transUnitWithId(ID_SOURCE_CHANGE);
            expect(tu).toBeTruthy();
            expect(tu.sourceContent()).toBe('Test Änderung Source (geändert!)');
            expect(tu.targetState()).toBe(dist_1.STATE_NEW);
            var tuFinal = langFileEnglish.transUnitWithId(ID_SOURCE_CHANGE_STATE_FINAL);
            expect(tuFinal).toBeTruthy();
            expect(tuFinal.sourceContent()).toBe('Test Änderung Source (state final, geändert!)');
            expect(tuFinal.targetState()).toBe(dist_1.STATE_TRANSLATED);
            done();
        });
        it('should merge changed descriptions and meanings to already translated files', function (done) {
            var TRANSLATED_FILE = 'WithDescriptions.en.xlf2';
            file_util_1.FileUtil.copy(SRCDIR + 'ngExtractedMasterWithDescriptions.xlf2', MASTER);
            file_util_1.FileUtil.copy(SRCDIR + TRANSLATED_FILE, WORKDIR + 'messages.en.xlf');
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'], verbose: true }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            expect(ws.writtenData()).toContain('WARNING: transferred 3 changed descriptions/meanings from master to "en"');
            // check that description and meaning are changed
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            var tu = langFileEnglish.transUnitWithId(ID_DESCRIPTION_CHANGE);
            expect(tu).toBeTruthy();
            expect(tu.description()).toBe('changed description');
            expect(tu.meaning()).toBe('changed meaning');
            // added description
            var tuAdded = langFileEnglish.transUnitWithId(ID_DESCRIPTION_ADD);
            expect(tuAdded).toBeTruthy();
            expect(tuAdded.description()).toBe('added description');
            expect(tuAdded.meaning()).toBe('added meaning');
            // removed description
            var tuRemoved = langFileEnglish.transUnitWithId(ID_DESCRIPTION_REMOVE);
            expect(tuRemoved).toBeTruthy();
            expect(tuRemoved.description()).toBeNull();
            expect(tuRemoved.meaning()).toBeNull();
            done();
        });
        it('should preserve order when merging new units (#96)', function (done) {
            file_util_1.FileUtil.copy(SRCDIR + 'preserveOrderMaster1.xlf2', MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'en',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['en', 'de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            // next step, use new master that has added 3 units
            file_util_1.FileUtil.copy(SRCDIR + 'preserveOrderMaster2.xlf2', MASTER);
            ws = new writer_to_string_1.WriterToString();
            commandOut = new command_output_1.CommandOutput(ws);
            xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['en', 'de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            expect(ws.writtenData()).toContain('merged 3 trans-units from master to "de"');
            // look, that the new file contains the new units at the correct position
            var langFileGerman = readXliff2(xliffMergeCmd.generatedI18nFile('de'));
            var addedTu = langFileGerman.transUnitWithId('addedunit1');
            expect(addedTu).toBeTruthy();
            expect(addedTu.sourceContent()).toBe('added unit 1');
            // check position
            expect(langFileGerman.editedContent().replace(/(\r\n|\n|\r)/gm, "")).toMatch(/addedunit1.*firstunit.*addedunit2.*lastunit.*addedunit3/);
            done();
        });
        it('should not remove trailing line break when merging', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var masterContent = file_util_1.FileUtil.read(MASTER, xml_reader_1.XmlReader.DEFAULT_ENCODING);
            expect(masterContent.endsWith('\n')).toBeTruthy('master file should end with EOL');
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var newContent = file_util_1.FileUtil.read(xliffMergeCmd.generatedI18nFile('de'), xml_reader_1.XmlReader.DEFAULT_ENCODING);
            expect(newContent.endsWith('\n')).toBeTruthy('file should end with EOL');
            done();
        });
        it('allowIdChange feature, should merge only white space changed content with changed ID to already translated files (#65)', function (done) {
            var ID_ORIGINAL = 'originalId';
            var ID_CHANGED = 'changedId';
            var TRANSLATED_FILE = 'WithIdChange.en.xlf2';
            file_util_1.FileUtil.copy(SRCDIR + 'ngExtractedMasterWithIdChange.xlf2', MASTER);
            file_util_1.FileUtil.copy(SRCDIR + TRANSLATED_FILE, WORKDIR + 'messages.en.xlf');
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    allowIdChange: true
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'], verbose: true }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            expect(ws.writtenData()).toContain('WARNING: found 2 changed id\'s in "en"');
            // check that changed id is merged
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            var tuChanged = langFileEnglish.transUnitWithId(ID_CHANGED);
            expect(tuChanged).toBeTruthy();
            expect(tuChanged.sourceContent().trim()).toBe('Test kleine Änderung, nur white spaces!');
            expect(tuChanged.targetState()).toBe(dist_1.STATE_TRANSLATED);
            expect(tuChanged.targetContent()).toBe('Test for a small white space change');
            var tuOriginal = langFileEnglish.transUnitWithId(ID_ORIGINAL);
            expect(tuOriginal).toBeFalsy();
            done();
        });
        it('allowIdChange feature, should merge untranslated only white space changed content with changed ID, but should set preserve state "new" (#68)', function (done) {
            var ID_ORIGINAL = 'originalIdUntranslated';
            var ID_CHANGED = 'changedIdUntranslated';
            var TRANSLATED_FILE = 'WithIdChange.en.xlf2';
            file_util_1.FileUtil.copy(SRCDIR + 'ngExtractedMasterWithIdChange.xlf2', MASTER);
            file_util_1.FileUtil.copy(SRCDIR + TRANSLATED_FILE, WORKDIR + 'messages.en.xlf');
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    allowIdChange: true
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'], verbose: true }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            expect(ws.writtenData()).toContain('WARNING: found 2 changed id\'s in "en"');
            // check that changed id is merged
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            var tuChanged = langFileEnglish.transUnitWithId(ID_CHANGED);
            expect(tuChanged).toBeTruthy();
            expect(tuChanged.sourceContent().trim()).toBe('Unübersetzt');
            expect(tuChanged.targetState()).toBe(dist_1.STATE_NEW);
            expect(tuChanged.targetContent()).toBe(''); // not translated
            var tuOriginal = langFileEnglish.transUnitWithId(ID_ORIGINAL);
            expect(tuOriginal).toBeFalsy();
            done();
        });
        it('should translate messages with placeholder', function (done) {
            file_util_1.FileUtil.copy(MASTER2SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de', 'en'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            // now translate some texts in the English version
            var langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            var tu = langFileEnglish.transUnitWithId(ID_WITH_PLACEHOLDER);
            expect(tu).toBeTruthy();
            tu.translate('Item <ph id="0" equiv="INTERPOLATION" disp="{{number()}}"/> of <ph id="1" equiv="INTERPOLATION_1" disp="{{total()}}"/> added.');
            translation_messages_file_reader_1.TranslationMessagesFileReader.save(langFileEnglish);
            // look, that the new file contains the translation
            langFileEnglish = readXliff2(xliffMergeCmd.generatedI18nFile('en'));
            expect(langFileEnglish.transUnitWithId(ID_WITH_PLACEHOLDER).targetContent()).toBe('Item <ph id="0" equiv="INTERPOLATION" disp="{{number()}}"/> of <ph id="1" equiv="INTERPOLATION_1" disp="{{total()}}"/> added.');
            expect(langFileEnglish.transUnitWithId(ID_WITH_PLACEHOLDER).targetContentNormalized().asDisplayString()).toBe('Item {{0}} of {{1}} added.');
            done();
        });
        it('should not output a warning when autotranslate is not enabled for a language (issue #49)', function (done) {
            file_util_1.FileUtil.copy(MASTER2SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var apiKey = auto_translate_service_spec_1.getApiKey();
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    languages: ['de', 'ru', 'en'],
                    autotranslate: ['ru'],
                    apikey: apiKey,
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, {}, profileContent);
            xliffMergeCmd.run(function () {
                if (apiKey) {
                    expect(ws.writtenData()).not.toContain('ERROR');
                    expect(ws.writtenData()).not.toContain('Auto translation from "de" to "en"');
                    expect(ws.writtenData()).toContain('Auto translation from "de" to "ru"');
                }
                else {
                    expect(ws.writtenData()).toContain('ERROR: autotranslate requires an API key');
                }
                done();
            });
        });
    });
    describe('ngx-translate processing for format XLIFF 2.0', function () {
        var MASTER1FILE = 'ngxtranslate.xlf2';
        var MASTER1SRC = SRCDIR + MASTER1FILE;
        var MASTER_WITHOUT_NGX_TRANSLATE_STUFF = SRCDIR + 'ngExtractedMaster1.xlf2';
        var MASTERFILE = 'messages.xlf2';
        var MASTER = WORKDIR + MASTERFILE;
        var ID_NODESC_NOMEANING = "2047558209369508311"; // an ID without set meaning and description
        var ID_MONDAY = "6830980354990918030"; // an ID from ngxtranslate.xlf with meaning "x.y" and description "ngx-translate"
        beforeEach(function () {
            if (!fs.existsSync(WORKDIR)) {
                fs.mkdirSync(WORKDIR);
            }
            // cleanup workdir
            file_util_1.FileUtil.deleteFolderContentRecursive(WORKDIR);
        });
        it('should return null for unset description and meaning in master xlf2 file', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var master = readXliff2(MASTER);
            expect(master.transUnitWithId(ID_NODESC_NOMEANING).description()).toBeFalsy();
            expect(master.transUnitWithId(ID_NODESC_NOMEANING).meaning()).toBeFalsy();
            done();
        });
        it('should find description and meaning in master xlf2 file', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var master = readXliff2(MASTER);
            expect(master.transUnitWithId(ID_MONDAY).description()).toBe('ngx-translate');
            expect(master.transUnitWithId(ID_MONDAY).meaning()).toBe('dateservice.monday');
            done();
        });
        it('should find description and meaning in translated xlf2 file', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var langFile = readXliff2(xliffMergeCmd.generatedI18nFile('de'));
            expect(langFile.transUnitWithId(ID_MONDAY).description()).toBe('ngx-translate');
            expect(langFile.transUnitWithId(ID_MONDAY).meaning()).toBe('dateservice.monday');
            done();
        });
        it('should write translation json file for ngx-translate', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    supportNgxTranslate: true
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var translationJsonFilename = xliffMergeCmd.generatedNgxTranslateFile('de');
            expect(file_util_1.FileUtil.exists(translationJsonFilename)).toBeTruthy();
            var fileContent = file_util_1.FileUtil.read(translationJsonFilename, 'UTF-8');
            var translation = JSON.parse(fileContent);
            expect(translation).toBeTruthy();
            expect(translation.myapp).toBeTruthy();
            expect(translation.dateservice.monday).toBe("Montag");
            expect(translation.dateservice.friday).toBe("Freitag");
            expect(translation.explicitlysetids.test1).toBe("Explizit gesetzte ID");
            done();
        });
        it('should handle placeholders in json file for ngx-translate', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    supportNgxTranslate: true
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var translationJsonFilename = xliffMergeCmd.generatedNgxTranslateFile('de');
            expect(file_util_1.FileUtil.exists(translationJsonFilename)).toBeTruthy();
            var fileContent = file_util_1.FileUtil.read(translationJsonFilename, 'UTF-8');
            var translation = JSON.parse(fileContent);
            expect(translation).toBeTruthy();
            expect(translation.placeholders).toBeTruthy();
            expect(translation.placeholders.test1placeholder).toBe('{{0}}: Eine Nachricht mit einem Platzhalter');
            expect(translation.placeholders.test2placeholder).toBe('{{0}}: Eine Nachricht mit 2 Platzhaltern: {{1}}');
            expect(translation.placeholders.test2placeholderRepeated).toBe('{{0}}: Eine Nachricht mit 2 Platzhaltern: {{0}} {{1}}');
            done();
        });
        it('should handle embedded html markup in json file for ngx-translate', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    supportNgxTranslate: true
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var translationJsonFilename = xliffMergeCmd.generatedNgxTranslateFile('de');
            expect(file_util_1.FileUtil.exists(translationJsonFilename)).toBeTruthy();
            var fileContent = file_util_1.FileUtil.read(translationJsonFilename, 'UTF-8');
            var translation = JSON.parse(fileContent);
            expect(translation).toBeTruthy();
            expect(translation.embeddedhtml).toBeTruthy();
            expect(translation.embeddedhtml.bold).toBe('Diese Nachricht ist <b>WICHTIG</b>');
            expect(translation.embeddedhtml.boldstrong).toBe('Diese Nachricht ist <b><strong>SEHR WICHTIG</strong></b>');
            expect(translation.embeddedhtml.strange).toBe('Diese Nachricht ist <strange>{{0}}</strange>');
            done();
        });
        it('should not export @@ids to translation json file, when this is supressed in pattern (issue #62)', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    supportNgxTranslate: true,
                    ngxTranslateExtractionPattern: 'ngx-translate'
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var translationJsonFilename = xliffMergeCmd.generatedNgxTranslateFile('de');
            expect(file_util_1.FileUtil.exists(translationJsonFilename)).toBeTruthy();
            var fileContent = file_util_1.FileUtil.read(translationJsonFilename, 'UTF-8');
            var translation = JSON.parse(fileContent);
            expect(translation).toBeTruthy();
            expect(translation.dateservice.monday).toBe("Montag");
            expect(translation.explicitlysetids).toBeFalsy();
            expect(translation["alt-ngx-translate"]).toBeFalsy();
            done();
        });
        it('should export other then ngx-translate description marked entries to translation json file, when this is specified in pattern (issue #62)', function (done) {
            file_util_1.FileUtil.copy(MASTER1SRC, MASTER);
            var ws = new writer_to_string_1.WriterToString();
            var commandOut = new command_output_1.CommandOutput(ws);
            var profileContent = {
                xliffmergeOptions: {
                    defaultLanguage: 'de',
                    srcDir: WORKDIR,
                    genDir: WORKDIR,
                    i18nFormat: 'xlf2',
                    i18nFile: MASTERFILE,
                    supportNgxTranslate: true,
                    ngxTranslateExtractionPattern: 'ngx-translate|alt-ngx-translate'
                }
            };
            var xliffMergeCmd = xliff_merge_1.XliffMerge.createFromOptions(commandOut, { languages: ['de'] }, profileContent);
            xliffMergeCmd.run();
            expect(ws.writtenData()).not.toContain('ERROR');
            var translationJsonFilename = xliffMergeCmd.generatedNgxTranslateFile('de');
            expect(file_util_1.FileUtil.exists(translationJsonFilename)).toBeTruthy();
            var fileContent = file_util_1.FileUtil.read(translationJsonFilename, 'UTF-8');
            var translation = JSON.parse(fileContent);
            expect(translation).toBeTruthy();
            expect(translation.dateservice.monday).toBe("Montag");
            expect(translation.explicitlysetids).toBeFalsy();
            expect(translation["alt-ngx-translate"].example1).toBe('Alternate description for ngx-translate export');
            done();
        });
    });
});
