// ngx-i18nsupport/dist/autotranslate/auto-translate-summary-report.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var util_1 = require("util");
/**
 * A report about a run of Google Translate over all untranslated unit.
 * * Created by martin on 29.06.2017.
 */
var AutoTranslateSummaryReport = /** @class */ (function () {
    function AutoTranslateSummaryReport(from, to) {
        this._from = from;
        this._to = to;
        this._total = 0;
        this._ignored = 0;
        this._success = 0;
        this._failed = 0;
    }
    /**
     * Set error if total call failed (e.g. "invalid api key" or "no connection" ...)
     * @param error
     * @param total
     */
    AutoTranslateSummaryReport.prototype.setError = function (error, total) {
        this._error = error;
        this._total = total;
        this._failed = total;
    };
    AutoTranslateSummaryReport.prototype.error = function () {
        return this._error;
    };
    AutoTranslateSummaryReport.prototype.setIgnored = function (ignored) {
        this._total += ignored;
        this._ignored = ignored;
    };
    /**
     * Add a single result to the summary.
     * @param tu
     * @param result
     */
    AutoTranslateSummaryReport.prototype.addSingleResult = function (tu, result) {
        // TODO
        this._total++;
        if (result.success()) {
            this._success++;
        }
        else {
            this._failed++;
        }
    };
    /**
     * Merge another summary into this one.
     * @param anotherSummary
     */
    AutoTranslateSummaryReport.prototype.merge = function (anotherSummary) {
        if (!this._error) {
            this._error = anotherSummary._error;
        }
        this._total += anotherSummary.total();
        this._ignored += anotherSummary.ignored();
        this._success += anotherSummary.success();
        this._failed += anotherSummary.failed();
    };
    AutoTranslateSummaryReport.prototype.total = function () {
        return this._total;
    };
    AutoTranslateSummaryReport.prototype.ignored = function () {
        return this._ignored;
    };
    AutoTranslateSummaryReport.prototype.success = function () {
        return this._success;
    };
    AutoTranslateSummaryReport.prototype.failed = function () {
        return this._failed;
    };
    /**
     * Human readable version of report
     */
    AutoTranslateSummaryReport.prototype.content = function () {
        var result;
        if (this._error) {
            result = util_1.format('Auto translation from "%s" to "%s" failed: "%s", failed units: %s', this._from, this._to, this._error, this._failed);
        }
        else {
            result = util_1.format('Auto translation from "%s" to "%s", total auto translated units: %s, ignored: %s, succesful: %s, failed: %s', this._from, this._to, this._total, this._ignored, this._success, this._failed);
        }
        return result;
    };
    return AutoTranslateSummaryReport;
}());
exports.AutoTranslateSummaryReport = AutoTranslateSummaryReport;
